src/v/kafka/server/group.cc

```diff
@@ -3218,4 +3222,102 @@ std::ostream& operator<<(std::ostream& o, const group::offset_metadata& md) {
     return o;
 }
 
+bool group::subscribed(const model::topic& topic) const {
+    if (_subscriptions.has_value()) {
+        return _subscriptions.value().contains(topic);
+    }
+    // answer conservatively
+    return true;
+}
+
+/*
+ * Currently only supports Version 0 which is a prefix version for higher
+ * versions and contains all of the information that we need.
+ */
+absl::node_hash_set<model::topic>
+group::decode_consumer_subscriptions(iobuf data) {
+    constexpr auto max_topic_name_length = 32_KiB;
+
+    request_reader reader(std::move(data));
+
+    /* version intentionally ignored */
+    reader.read_int16();
+
+    const auto count = reader.read_int32();
+    if (count < 0) {
+        throw std::out_of_range(fmt::format(
+          "consumer metadata contains negative topic count {}", count));
+    }
+
+    // a simple heuristic to avoid large allocations
+    if (static_cast<size_t>(count) > reader.bytes_left()) {
+        throw std::out_of_range(fmt::format(
+          "consumer metadata topic count too large {} > {}",
+          count,
+          reader.bytes_left()));
+    }
+
+    absl::node_hash_set<model::topic> topics;
+    topics.reserve(count);
+
+    while (topics.size() != static_cast<size_t>(count)) {
+        const auto len = reader.read_int16();
+        if (len < 0) {
+            throw std::out_of_range(fmt::format(
+              "consumer metadata contains negative topic name length {}", len));
+        } else if (static_cast<size_t>(len) > max_topic_name_length) {
+            throw std::out_of_range(fmt::format(
+              "consumer metadata contains topic name that exceeds maximum size "
+              "{} > {}",
+              len,
+              max_topic_name_length));
+        }
+        auto name = reader.read_string_unchecked(len);
+        topics.insert(model::topic(std::move(name)));
+    }
+
+    // after the topic list is user data bytes, followed by extra bytes for
+    // versions > 0, neither of which are needed.
+
+    return topics;
+}
+
+void group::update_subscriptions() {
+    if (_protocol_type != "consumer") {
```

Comment:
Maybe these can be enums `protocol_type::consumer`, if its not too much work to do

Reply:
the protocol names are actually free-form strings provided by kafka clients :( the "consumer" one is the only well-known name.